### PR TITLE
Add loopback chain-element

### LINKED ISSUE #364: Create sdk-vpp chain element for creating loopback interface attached to the fib table
## Depends on
https://github.com/networkservicemesh/sdk-vpp/issues/341

## Context
https://github.com/networkservicemesh/cmd-nse-vl3-vpp/issues/1

## Description

We need to 
1. create a loopback interface attached to the fib table
2. assign its IP address to the loopback interface.
3.  program a route in the fib table to the SrcIP for the NSC connecting and point out the interface.

### PR DESCRIPTION
**Pull request by glazychev-art**
Closes #364 

Signed-off-by: Artem Glazychev <artem.glazychev@xored.com>

### CODE REVIEW

File: pkg/networkservice/loopback/server.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+			entries: make(map[string]*loopInfo),
+		},
+	}
+}
+
+func (l *loopback) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*networkservice.Connection, error) {
+	if err := createLoopback(ctx, request.GetConnection(), l.vppConn, &l.loopbacks, metadata.IsClient(l)); err != nil {
+		return nil, err
+	}
+	return next.Server(ctx).Request(ctx, request)
```
Comment (Bolodya1997) [inline pkg/networkservice/loopback/server.go:51]: There will be no timeout or Close if it is an initial Request and it fails further in the chain. Please add some cleanup for such case.

Comment (edwarnicke) [reply pkg/networkservice/loopback/server.go:51]: @artem-belov Could you respond to this comment?

Comment (glazychev-art) [reply pkg/networkservice/loopback/server.go]: Fixed

File: pkg/networkservice/loopback/option.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+type LoopMap struct {
+	/* entries - is a map[NetworkServiceName]{swIfIndex, count} */
+	entries map[string]*loopInfo
+
+	/* mutex for entries */
+	mut sync.Mutex
+}
+
+// CreateLoopbackMap creates loopback map
+func CreateLoopbackMap() *LoopMap {
```
Comment (denis-tingaikin) [inline pkg/networkservice/loopback/option.go:43]:
```suggestion
func NewMap() *LoopMap {
```

**Review by denis-tingaikin — APPROVED** [review]
Added a few minor comments. Looks good.
@edwarnicke Could you have a look?

File: pkg/networkservice/loopback/option.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+type loopInfo struct {
+	/* loopback swIfIndex */
+	swIfIndex interface_types.InterfaceIndex
+
+	/* count - the number of clients using this loopback */
+	count uint32
+}
+
+// LoopMap stores loopback swIfIndex by NetworkServiceName
+type LoopMap struct {
```
Comment (denis-tingaikin) [inline pkg/networkservice/loopback/option.go:34]: It's already in loopback pkg.
```suggestion
type Map struct {
```

File: pkg/networkservice/loopback/option.go (hunk 2 above)
Comment (glazychev-art) [reply pkg/networkservice/loopback/option.go]: fixed

File: pkg/networkservice/loopback/option.go (hunk 3 above)
Comment (glazychev-art) [reply pkg/networkservice/loopback/option.go]: fixed

**Review by Mixaster995 — APPROVED** [review]

File: pkg/networkservice/loopback/option.go [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+	count uint32
+}
+
+// Map stores loopback swIfIndex by NetworkServiceName
+type Map struct {
+	/* entries - is a map[NetworkServiceName]{swIfIndex, count} */
+	entries map[string]*loopInfo
+
+	/* mutex for entries */
+	mut sync.Mutex
```
Comment (edwarnicke) [inline pkg/networkservice/loopback/option.go:39]: I'm concerned about having a global mutex here that will serialize *all* requests behind it.  Is there some combination of something like sync.Map we could use to reduce the contention?

Comment (glazychev-art) [reply pkg/networkservice/loopback/option.go]: @edwarnicke 
I think the only one better way is to use `executor` instead of `mutex`
Fixed

**Review by Mixaster995 — APPROVED** [review]

File: pkg/networkservice/loopback/option.go [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+type Map struct {
+	/* entries - is a map[NetworkServiceName]{swIfIndex, count} */
+	entries map[string]*loopInfo
+
+	/* executor */
+	exec serialize.Executor
+}
+
+// NewMap creates loopback map
+func NewMap() *Map {
```
Comment (denis-tingaikin) [inline pkg/networkservice/loopback/option.go:42]: Can be private.

```suggestion
func newMap() *Map {
```

**Review by denis-tingaikin — APPROVED** [review]
LGTM

File: pkg/networkservice/loopback/option.go (hunk 5 above)
Comment (denis-tingaikin) [reply pkg/networkservice/loopback/option.go:42]: OK, it will be used in other repos

**Comment by denis-tingaikin** [comment]
@edwarnicke Can we merge this? Do you have any comments?

**Review by edwarnicke — APPROVED** [review]
LGTM, but needs a rebase

**Pull request closed without merging**